lib/indexer.rb

```diff
@@ -98,18 +101,25 @@ def find_cdxj_files
   end
 
   def write_cdxj(warc_file)
-    cdxj_path = cdxj_path(warc_file)
-    cdxj_path.parent.mkpath unless cdxj_path.parent.directory?
-
-    cmd = ['cdxj-indexer', '--post-append', '--output', cdxj_path.to_s, warc_file.to_s]
-    @log.info("generating cdxj #{cdxj_path} for #{warc_file}")
-    @log.error("cdxj error while parsing #{warc}") unless system(*cmd)
+    cdxj = cdxj_path(warc_file)
+    cdxj.parent.mkpath unless cdxj.parent.directory?
+    cdxj_temp = Pathname.new(cdxj.to_s + '.tmp')
+
+    cmd = ['cdxj-indexer', '--dir-root', @warcs_dir.to_s, '--post-append', '--output', cdxj_temp.to_s, warc_file.to_s]
+    @log.info("generating cdxj #{cdxj} for #{warc_file}")
+    unless system(*cmd)
+      @log.error("cdxj error while parsing #{warc_file}")
+      cdxj_temp.unlink
+      return
+    end
 
-    cdxj_path
+    # guard against the index having been already generated by another process
+    cdxj_temp.rename(cdxj) if cdxj_temp.file?
+    cdxj
   end
 
   def cdxj_path(warc_file)
-    cdx_filename = warc_file.basename.to_s.sub(/w?arc.gz$/, 'cdxj')
+    cdx_filename = warc_file.basename.to_s.sub(/w?arc(.gz)?$/, 'cdxj')
```

Comment:
Not all WARC files are compressed, this makes sure that their temporary CDXJ file is named with a `.cdxj` extension  rather than `.warc`.